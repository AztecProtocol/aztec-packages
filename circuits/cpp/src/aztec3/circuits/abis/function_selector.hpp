#pragma once

#include "aztec3/utils/types/circuit_types.hpp"
#include "aztec3/utils/types/convert.hpp"
#include "aztec3/utils/types/native_types.hpp"

#include "barretenberg/serialize/msgpack.hpp"

namespace aztec3::circuits::abis {

using aztec3::utils::types::CircuitTypes;
using aztec3::utils::types::NativeTypes;

template <typename NCT> struct FunctionSelector {
    using uint32 = typename NCT::uint32;
    using boolean = typename NCT::boolean;
    using fr = typename NCT::fr;

    uint32 value;  // e.g. 1st 4-bytes of abi-encoding of function.

    MSGPACK_FIELDS(value);

    boolean operator==(FunctionSelector<NCT> const& other) const { return value == other.value; };

    template <typename Builder> FunctionSelector<CircuitTypes<Builder>> to_circuit_type(Builder& builder) const
    {
        static_assert((std::is_same<NativeTypes, NCT>::value));

        // Capture the circuit builder:
        auto to_ct = [&](auto& e) { return aztec3::utils::types::to_ct(builder, e); };

        FunctionSelector<CircuitTypes<Builder>> function_selector = {
            to_ct(value),
        };

        return function_selector;
    };

    template <typename Builder> FunctionSelector<NativeTypes> to_native_type() const
    {
        static_assert(std::is_same<CircuitTypes<Builder>, NCT>::value);
        auto to_nt = [&](auto& e) { return aztec3::utils::types::to_nt<Builder>(e); };

        FunctionSelector<NativeTypes> function_selector = {
            to_nt(value),
        };

        return function_selector;
    };

    void set_public()
    {
        static_assert(!(std::is_same<NativeTypes, NCT>::value));

        fr(value).set_public();
    }
};

}  // namespace aztec3::circuits::abis