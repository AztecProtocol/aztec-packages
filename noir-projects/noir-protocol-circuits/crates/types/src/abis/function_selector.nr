use crate::traits::{Deserialize, Empty, FromField, Serialize, ToField};

pub struct FunctionSelector {
    // 1st 4-bytes of abi-encoding of function.
    inner: u32,
}

impl Eq for FunctionSelector {
    fn eq(self, function_selector: FunctionSelector) -> bool {
        function_selector.inner == self.inner
    }
}

impl Serialize<1> for FunctionSelector {
    fn serialize(self: Self) -> [Field; 1] {
        [self.inner as Field]
    }
}

impl Deserialize<1> for FunctionSelector {
    fn deserialize(fields: [Field; 1]) -> Self {
        Self { inner: fields[0] as u32 }
    }
}

impl FromField for FunctionSelector {
    fn from_field(field: Field) -> Self {
        Self { inner: field as u32 }
    }
}

impl ToField for FunctionSelector {
    fn to_field(self) -> Field {
        self.inner as Field
    }
}

impl Empty for FunctionSelector {
    fn empty() -> Self {
        Self { inner: 0 as u32 }
    }
}

impl FunctionSelector {
    pub fn from_u32(value: u32) -> Self {
        Self { inner: value }
    }

    pub fn from_signature<let N: u32>(signature: str<N>) -> Self {
        let bytes = signature.as_bytes();
        let hash = crate::hash::poseidon2_hash_bytes(bytes);

        // `hash` is automatically truncated to fit within 32 bits.
        FunctionSelector::from_field(hash)
    }

    pub fn zero() -> Self {
        Self { inner: 0 }
    }
}

#[test]
fn test_is_valid_selector() {
    let selector = FunctionSelector::from_signature("IS_VALID()");
    assert_eq(selector.to_field(), 0x73cdda47);
}

#[test]
fn test_long_selector() {
    let selector =
        FunctionSelector::from_signature("foo_and_bar_and_baz_and_foo_bar_baz_and_bar_foo");
    assert_eq(selector.to_field(), 0x7590a997);
}
