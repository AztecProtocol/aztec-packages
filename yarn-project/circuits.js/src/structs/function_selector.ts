import { ABIParameter } from '@aztec/foundation/abi';
import { keccak } from '@aztec/foundation/crypto';
import { BufferReader } from '@aztec/foundation/serialize';

import { FUNCTION_SELECTOR_NUM_BYTES } from '../cbind/constants.gen.js';

/**
 * A function selector is the first 4 bytes of the hash of a function signature.
 */
export class FunctionSelector {
  /**
   * The size of the hash in bytes.
   */
  public static SIZE = FUNCTION_SELECTOR_NUM_BYTES;

  constructor(/** buffer containing the function selector */ public value: Buffer) {
    if (value.length !== FunctionSelector.SIZE) {
      throw new Error(`Function selector must be ${FunctionSelector.SIZE} bytes long, got ${value.length} bytes.`);
    }
  }

  /**
   * Checks if the function selector is empty (all bytes are 0).
   * @returns True if the function selector is empty (all bytes are 0).
   */
  public isEmpty(): boolean {
    return this.value.equals(Buffer.alloc(FunctionSelector.SIZE));
  }

  /**
   * Serialize as a buffer.
   * @returns The buffer.
   */
  toBuffer(): Buffer {
    return this.value;
  }

  /**
   * Deserializes from a buffer or reader, corresponding to a write in cpp.
   * @param buffer - Buffer  or BufferReader to read from.
   * @returns The FunctionSelector.
   */
  static fromBuffer(buffer: Buffer | BufferReader): FunctionSelector {
    const reader = BufferReader.asReader(buffer);
    return new FunctionSelector(reader.readBytes(FunctionSelector.SIZE));
  }

  /**
   * Creates a function selector from a signature.
   * @param signature - Signature of the function to generate the selector for (e.g. "transfer(field,field)").
   * @returns Function selector.
   */
  static fromSignature(signature: string): FunctionSelector {
    return new FunctionSelector(keccak(Buffer.from(signature)).subarray(0, FunctionSelector.SIZE));
  }

  /**
   * Creates a function selector for a given function name and parameters.
   * @param name - The name of the function.
   * @param parameters - An array of ABIParameter objects, each containing the type information of a function parameter.
   * @returns A Buffer containing the 4-byte function selector.
   */
  static fromNameAndParameters(name: string, parameters: ABIParameter[]) {
    const signature = name === 'constructor' ? name : `${name}(${parameters.map(p => p.type.kind).join(',')})`;
    return FunctionSelector.fromSignature(signature);
  }

  /**
   * Creates an empty function selector.
   * @returns An empty function selector.
   */
  static empty(): FunctionSelector {
    return new FunctionSelector(Buffer.alloc(FunctionSelector.SIZE));
  }
}
