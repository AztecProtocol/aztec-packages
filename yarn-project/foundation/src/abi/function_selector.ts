import { ABIParameter, decodeFunctionSignature } from '@aztec/foundation/abi';
import { toBigIntBE, toBufferBE } from '@aztec/foundation/bigint-buffer';
import { keccak } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';

/**
 * A function selector is the first 4 bytes of the hash of a function signature.
 */
export class FunctionSelector {
  /**
   * The size of the function selector in bytes.
   */
  public static SIZE = 4;

  constructor(/** number representing the function selector */ public value: number) {
    if (value > 2 ** (FunctionSelector.SIZE * 8) - 1) {
      throw new Error(`Function selector must fit in ${FunctionSelector.SIZE} bytes.`);
    }
  }

  /**
   * Checks if the function selector is empty (all bytes are 0).
   * @returns True if the function selector is empty (all bytes are 0).
   */
  public isEmpty(): boolean {
    return this.value === 0;
  }

  /**
   * Serialize as a buffer.
   * @returns The buffer.
   */
  toBuffer(): Buffer {
    return toBufferBE(BigInt(this.value), FunctionSelector.SIZE);
  }

  /**
   * Serialize as a hex string.
   * @returns The string.
   */
  toString(): string {
    return this.toBuffer().toString('hex');
  }

  /**
   * Checks if this function selector is equal to another.
   * @param other - The other function selector.
   * @returns True if the function selectors are equal.
   */
  equals(other: FunctionSelector): boolean {
    return this.value === other.value;
  }

  /**
   * Deserializes from a buffer or reader, corresponding to a write in cpp.
   * @param buffer - Buffer  or BufferReader to read from.
   * @returns The FunctionSelector.
   */
  static fromBuffer(buffer: Buffer | BufferReader): FunctionSelector {
    const reader = BufferReader.asReader(buffer);
    const value = Number(toBigIntBE(reader.readBytes(FunctionSelector.SIZE)));
    return new FunctionSelector(value);
  }

  /**
   * Returns a new field with the same contents as this EthAddress.
   *
   * @returns An Fr instance.
   */
  public toField() {
    return new Fr(this.value);
  }

  /**
   * Converts a field to function selector.
   * @param fr - The field to convert.
   * @returns The function selector.
   */
  static fromField(fr: Fr): FunctionSelector {
    return new FunctionSelector(Number(fr.value));
  }

  /**
   * Creates a function selector from a signature.
   * @param signature - Signature of the function to generate the selector for (e.g. "transfer(field,field)").
   * @returns Function selector.
   */
  static fromSignature(signature: string): FunctionSelector {
    return FunctionSelector.fromBuffer(keccak(Buffer.from(signature)).subarray(0, FunctionSelector.SIZE));
  }

  /**
   * Creates a function selector for a given function name and parameters.
   * @param name - The name of the function.
   * @param parameters - An array of ABIParameter objects, each containing the type information of a function parameter.
   * @returns A Buffer containing the 4-byte function selector.
   */
  static fromNameAndParameters(name: string, parameters: ABIParameter[]) {
    const signature = decodeFunctionSignature(name, parameters);
    const selector = FunctionSelector.fromSignature(signature);
    // If using the debug logger here it kill the typing in the `server_world_state_synchroniser` and jest tests.
    // console.log(`Function selector for ${signature} is ${selector}`);
    return selector;
  }

  /**
   * Creates an empty function selector.
   * @returns An empty function selector.
   */
  static empty(): FunctionSelector {
    return new FunctionSelector(0);
  }
}
