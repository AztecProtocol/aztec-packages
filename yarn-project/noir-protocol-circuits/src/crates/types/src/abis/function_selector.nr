struct FunctionSelector {
    // 1st 4-bytes of abi-encoding of function.
    inner : u32,
}

impl FunctionSelector {
    fn to_field(self) -> Field {
        self.inner as Field
    }

    pub fn from_u32(value : u32) -> FunctionSelector {
        FunctionSelector {
            inner : value,
        }
    }

    pub fn zero() -> Self {
        FunctionSelector { inner: 0 }
    }
}
