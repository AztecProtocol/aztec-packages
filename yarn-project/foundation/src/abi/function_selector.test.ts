import { FunctionSelector } from './function_selector.js';

describe('FunctionSelector', () => {
  let selector: FunctionSelector;

  beforeAll(() => {
    selector = FunctionSelector.random();
  });

  it('serializes to buffer and deserializes it back', () => {
    const buffer = selector.toBuffer();
    const res = FunctionSelector.fromBuffer(buffer);
    expect(res).toEqual(selector);
    expect(res.isEmpty()).toBe(false);
  });

  it('serializes to field and deserializes it back', () => {
    const field = selector.toField();
    const res = FunctionSelector.fromField(field);
    expect(res).toEqual(selector);
  });

  it('computes a function selector from signature', () => {
    const res = FunctionSelector.fromSignature('transfer(address,uint256)');
    expect(res).toMatchSnapshot();
  });
});
