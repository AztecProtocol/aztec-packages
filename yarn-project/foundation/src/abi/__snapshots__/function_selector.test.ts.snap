// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FunctionSelector computes a function selector from signature 1`] = `
FunctionSelector {
  "value": 2835717307,
}
`;
